### ISSUE TITLE
Slow training

### ISSUE THREAD

**Issue by jeanollion** [kaloeffler/EmbedTrack#1]
Hello, 
I would like to try you method on my data. 
I formatted the data as in CTC and run the notebook, using the default parameters. Training is extremely slow (18s / iteration, 6h / epoch), even-though the GPU (NVIDIA RTX A6000) is being used. 
My dataset is composed of very dense bacteria (~300-400 bacteria per 256x256 patch), could this explain the slow training speed ? Or is there something else I could do to speed up the training ? 
Thanks in advance.

**Comment by kaloeffler** [comment]
Dear Jean,

that's quite hard to tell from afar, what might cause this issue. Depending on the GPU usage you might test increasing batch size / number of workers in the dataloader or decreasing the batch size.
1) Are you training on colab / or locally on your PC / using AWS ...? 

2) How large is your training dataset in total? Is just the validation slow or both?

Unfortunately, we never had access to that dense type of cell images, hence there might be also some scaling issues involved we are currently unaware of.  

3) Could you run some profiling (https://pytorch.org/docs/stable/profiler.html)? This would help a lot in where to look!

**Comment by jeanollion** [comment]
Dear Katharina, 
Thank you for your answer. 
GPU is clearly underused. To modify the number of workers, is there another way than editing the function `init_training` ? 
1. I'm running locally on my PC
2. The output of the dataloader says there are 19260 cropped images in the training set and 2140 in the validation set, so 1197 training steps and 127 validation steps. Both training and validation are slow: 18s / iteration for training and 9 s/ iteration for validation. 

What is supposed to be the training iteration time for the CTC dataset <em>Fluo-N2DH-SIM+</em> ? I will run on this dataset to check if there is an issue with my setup or if it is related to my dataset.

**Comment by kaloeffler** [comment]
Dear Jean,

I went again through the code and may have found the likely culprit. A part of the loss needs to be calculated for each instance in an image which is implemented by looping over them. Hence for very densely packed patches the calculation will be slow. Unfortunately I currently don't see a way to speed up this part of the code. 

However, during our training runs we observed that not excessive amounts of training data are required. Thus one approach to at least try whether the approach might be working for you is reducing the number of image patches per epoch which can be simply set in the script by setting a small number of samples for ` TRAIN_SIZE` and `VAL_SIZE` -- which will result in randomly selecting the number of samples during train / eval from your overall amount of train/val samples. I would suggest starting very small (~2% of the size of your current train and val data) and check whether the model yields acceptable results.

This is the part of the .ipynb which needs to be adapted:
```
TRAIN_SIZE = None  # if None training on full train data set; otherwise still training on full data set but only use a fraction of the data per epoch
VAL_SIZE = None  # if None validation on full val data set; otherwise still val on full data set but only use a fraction of the data per epoch
```

**Comment by jeanollion** [comment]
Dear Katharina, 
I let the training run until the end for a few days on the whole dataset. Your method  works on the data, with some errors though, but this was expected as it is a quite difficult dataset. 
Thank you for your answers. 
Best, 
Jean
